### PR TITLE
Use Jenkins design library icons

### PR DESCRIPTION
**Pull request by cronik**
Use Jenkins design library symbols and colors to favorite actions.

Updated to Jenkins 2.361.4 baseline and includes a number of simplifications and fixes.

- Use User.current() to check if user is not anonymous
- Fix bug where FavoriteListener events were never fired
- Implement ListViewColumnDescriptor#showByDefault to optimize column rendering (only needs be called once instead of for each row)
- Remove per row script block in favor of toggleFavoriteOnClick assets.js function
- Add integration tests

# Screenshots
* Favorite active action 
  Screenshot 2023-02-12 at 3 47 39 PM
* Favorite inactive action 
   Screenshot 2023-02-12 at 3 47 51 PM
* List view 
   Screenshot 2023-02-12 at 3 48 13 PM

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by cronik** [comment]
@olamy / @alecharp bumping the jenkins version caused the java 8 build to fail. The main Jenkinsfile probably needs to be adjusted by a maintainer.

File: pom.xml [hunk 1]
```diff
@@ -97,6 +99,16 @@
...
     </dependency>
+    <dependency>
+      <groupId>io.jenkins.plugins</groupId>
+      <artifactId>ionicons-api</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.jenkins-ci.plugins</groupId>
+      <artifactId>bouncycastle-api</artifactId>
+      <version>2.27</version>
+      <scope>test</scope>
```
Comment (olamy) [inline pom.xml:110]: why this new dependency?

File: pom.xml [hunk 2]
```diff
@@ -32,6 +32,8 @@
     <gitHubRepo>jenkinsci/${project.artifactId}-plugin</gitHubRepo>
     <!-- security spotbugs -->
     <spotbugs.failOnError>false</spotbugs.failOnError>
+    <spotbugs.effort>Max</spotbugs.effort>
+    <spotbugs.threshold>Low</spotbugs.threshold>
```
Comment (olamy) [inline pom.xml:36]: doesn't look related to the title "Use Jenkins design library icons"

File: src/main/java/hudson/plugins/favorite/FavoritePlugin.java [hunk 3]
```diff
@@ -7,69 +7,35 @@
...
-    public void doToggleFavorite(StaplerRequest req, StaplerResponse resp, @QueryParameter String job, @QueryParameter Boolean redirect) throws IOException {
-        Jenkins jenkins = Jenkins.get();
-        if (jenkins == null) {
-            throw new IllegalStateException("Jenkins not started");
-        }
+
+    @SuppressWarnings(value = "unused") // used by stapler
+    public void doToggleFavorite(StaplerRequest req, StaplerResponse resp, @QueryParameter String job, @QueryParameter Boolean redirect) throws IOException, FavoriteException, ServletException {
         User user = User.current();
-        if (user != null && !isAnonymous(user)) {
```
Comment (olamy) [inline src/main/java/hudson/plugins/favorite/FavoritePlugin.java:27]: doesn't look related to the title "Use Jenkins design library icons"

File: src/main/java/hudson/plugins/favorite/user/FavoriteUserProperty.java [hunk 4]
```diff
@@ -40,38 +41,58 @@ public FavoriteUserProperty() {}
...
+        return List.copyOf(getAllFavorites());
     }
 
+    /**
+     * Get all favorite jobs.
+     * @return unmodifiable favorites job set
+     */
     public Set<String> getAllFavorites() {
         removeFavoritesWhichDoNotExist();
-        return Maps.filterEntries( data, input -> input != null && input.getValue() ).keySet();
```
Comment (olamy) [inline src/main/java/hudson/plugins/favorite/user/FavoriteUserProperty.java:61]: doesn't look related to the title "Use Jenkins design library icons"

**Review by olamy — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution and sorry for the late review.
A lot of changes doesn't look related to the title "Use Jenkins design library icons"
they should be part of different.
This makes review easy and eventually revert easier.

File: src/main/java/hudson/plugins/favorite/Favorites.java [hunk 5]
```diff
@@ -150,7 +152,7 @@ public void remove() {
      * @throws IllegalArgumentException when the user is anonymous
      */
     private static FavoriteUserProperty getProperty(@NonNull User user) throws IOException {
-        if (FavoritePlugin.isAnonymous(user)) {
+        if ("anonymous".equalsIgnoreCase(user.getFullName())) {
```
Comment (mawinter69) [inline src/main/java/hudson/plugins/favorite/Favorites.java:155]: That is wrong. If `user` is not null, it means that the user is logged in and is not anonymous.

File: src/main/java/hudson/plugins/favorite/filter/FavoriteFilter.java [hunk 6]
```diff
@@ -23,15 +22,10 @@ public FavoriteFilter() {
     @Override
     public List<TopLevelItem> filter(List<TopLevelItem> added, List<TopLevelItem> all, View filteringView) {
         List<TopLevelItem> filtered = new ArrayList<>();
-
-        Authentication authentication = Hudson.getAuthentication();
-
+        Authentication authentication = Jenkins.getAuthentication2();
```
Comment (mawinter69) [inline src/main/java/hudson/plugins/favorite/filter/FavoriteFilter.java:25]: To be consistent with all other places I would suggest to use `User.current()` which already ensures that when not null, the user is logged in and when null, access happens as anonymous.

File: src/main/java/hudson/plugins/favorite/user/FavoriteUserProperty.java [hunk 7]
```diff
@@ -81,16 +102,36 @@ public boolean toggleFavorite(String job) throws IOException {
...
     }
 
     @Override
-    public UserProperty reconfigure(StaplerRequest req, JSONObject form) throws Descriptor.FormException {
+    public UserProperty reconfigure(StaplerRequest req, JSONObject form) {
+        if (form != null) {
+            for (Object key : form.keySet()) {
+                String job = key.toString();
+                if (!form.getBoolean(job)) {
+                    data.remove(job);
```
Comment (mawinter69) [inline src/main/java/hudson/plugins/favorite/user/FavoriteUserProperty.java:116]: This changes the behaviour of the plugin. Previously a click on the star immediately toggled the state. Now you have to explicitly save your properties.
Also the old behaviour was doing here `data.put(job, false)` and not remove the entry from the map.

File: src/main/resources/hudson/plugins/favorite/column/FavoriteColumn/column.jelly [hunk 8]
```diff
@@ -1,27 +1,16 @@
...
-      <a href="#">
-        <span id="fav_${job.getFullName()}" data-fullName="${job.getFullName()}"
-              class="${starClass} ${sizeClass}"/>
+<j:jelly xmlns:j="jelly:core" xmlns:l="/lib/layout">
+  <j:set var="sortData" value="${it.getSortData(job.getFullName())}"/>
+  <j:set var="fav" value="${it.isFavorite(job.getFullName())}"/>
+  <j:set var="sizeClass" value="${icons.toNormalizedIconSizeClass(subIconSize)}"/>
+  <td data="${sortData}" class="jenkins-table__cell--tight">
+    <div class="jenkins-table__cell__button-wrapper">
+      <a href="#" class="jenkins-table__button" onclick="toggleFavoriteOnClick(event, this)" data-fullName="${job.getFullName()}">
```
Comment (mawinter69) [inline src/main/resources/hudson/plugins/favorite/column/FavoriteColumn/column.jelly:8]: This uses inline java code which violates CSP. The onclick behaviour should be attached within the assets.js with `Behaviour.specify(...)`

File: src/main/resources/hudson/plugins/favorite/column/FavoriteColumn/column.jelly [hunk 9]
```diff
@@ -1,27 +1,16 @@
...
+        <span id="fav_${job.getFullName()}" data-fav="${fav}">
+            <l:icon src="symbol-star plugin-ionicons-api" class="icon-fav-active jenkins-!-color-yellow ${sizeClass} ${!fav ? 'jenkins-hidden' : ''}"/>
```
Comment (mawinter69) [inline src/main/resources/hudson/plugins/favorite/column/FavoriteColumn/column.jelly:10]: Will that potentially cause some flickering as you hide on icon and show another?
Maybe always use the `symbol-star` and just change the color in js by adding removing the class `jenkins-!-color-yellow`

File: src/main/resources/hudson/plugins/favorite/user/FavoriteUserProperty/config.jelly [hunk 10]
```diff
@@ -1,33 +1,12 @@
 <?jelly escape-by-default='true'?>
-<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout"
-         xmlns:t="/lib/hudson" xmlns:f="/lib/form">
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:f="/lib/form">
   <st:adjunct includes="hudson.plugins.favorite.assets"/>
```
Comment (mawinter69) [inline src/main/resources/hudson/plugins/favorite/user/FavoriteUserProperty/config.jelly:3]: With changing this to checkboxes, the inclusion of the js is no longer necessary I think.

**Comment by cronik** [comment]
Trimmed this PR down and will submit additional PRs to address other issues. The integration test that was added has some `TODO`'s as a result.

@mawinter69 I see you already created #141 for the listener issue. There is a commented out assertion on FavoritesPluginTest line 95 depending on which PR goes first.

File: src/main/resources/hudson/plugins/favorite/column/FavoriteColumn/column.jelly (hunk 9 above)
Comment (cronik) [reply src/main/resources/hudson/plugins/favorite/column/FavoriteColumn/column.jelly:10]: There is no noticeable flickering that I've observed. I originally has the solid star and changed the color but after testing it was unclear that the non-yellow star meant that it was not toggled. The empty star makes the state more clear in my opinion.

File: src/main/resources/hudson/plugins/favorite/column/FavoriteColumn/column.jelly (hunk 8 above)
Comment (cronik) [reply src/main/resources/hudson/plugins/favorite/column/FavoriteColumn/column.jelly]: I assume the inline java code you are referencing is `data-fullName="${job.getFullName()}"`. Shouldn't `<?jelly escape-by-default='true'?>` at the top of the jelly file mitigate that? If not can you suggest a remedy.

Moved `onclick` to `Behaviour.specify` ✅

File: src/main/java/hudson/plugins/favorite/Favorites.java (hunk 5 above)
Comment (cronik) [reply src/main/java/hudson/plugins/favorite/Favorites.java]: You may be right, I was just keeping the logic that was there before. If that is indeed the case then there is no need for the anonymous check when user is declared as `@NonNull`. I can clean that up in a future PR.

File: src/main/java/hudson/plugins/favorite/user/FavoriteUserProperty.java (hunk 7 above)
Comment (cronik) [reply src/main/java/hudson/plugins/favorite/user/FavoriteUserProperty.java]: Reverted to original implementation

**Review by mawinter69 — CHANGES_REQUESTED** [review]

File: src/main/resources/hudson/plugins/favorite/project/FavoriteProjectAction/action.jelly [hunk 11]
```diff
@@ -1,10 +1,12 @@
...
-            permission="${app.READ}"
-            contextMenu="false"
-            post="true"
-    />
+    <j:if test="${icon!=null}">
+        <l:task icon="${icon}" title="${action.displayName}"
+                href="${rootURL}/${action.urlName}"
+                permission="${app.READ}"
+                contextMenu="false"
+                post="false"
```
[the remark this answers is not in the document]
Comment (cronik) [reply src/main/resources/hudson/plugins/favorite/project/FavoriteProjectAction/action.jelly]: The reason for not using post is due to the implementation of `l:task`. It does not allow icon class attributes or the ability to switch icons. Didn't see any downside to performing a get and redirect as also figured using the `l:task` tag would be better long term in case the task/action ui classes and structure changes in the future. For all those reasons this seemed like the best compromise of function vs unnecessary complexity. 

Re-worked with your suggestions.

**Review by mawinter69 — APPROVED** [review]

**Comment by NotMyFault** [comment]
While this PR is a great addition for the plugin, it broke the acceptance test suite, unfortunately. Specifically, the removal of `fav_` divs:
```
[2023-12-12T18:31:32.537Z] [ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 302.3 s <<< FAILURE! -- in plugins.FavoriteTest
[2023-12-12T18:31:32.537Z] [ERROR] plugins.FavoriteTest.shouldBeAbleToFavoriteAJob -- Time elapsed: 152.3 s <<< ERROR!
[2023-12-12T18:31:32.537Z] org.openqa.selenium.TimeoutException: Expected condition failed: Element matching By.cssSelector: #fav_my-project is present (tried for 120 second(s) with 500 milliseconds interval)
[2023-12-12T18:31:32.537Z] 	at org.openqa.selenium.support.ui.FluentWait.timeoutException(FluentWait.java:262)
[2023-12-12T18:31:32.537Z] 	at org.jenkinsci.test.acceptance.junit.Wait.timeoutException(Wait.java:188)
[2023-12-12T18:31:32.537Z] 	at org.openqa.selenium.support.ui.FluentWait.until(FluentWait.java:230)
[2023-12-12T18:31:32.537Z] 	at org.jenkinsci.test.acceptance.junit.Wait.until(Wait.java:119)
[2023-12-12T18:31:32.537Z] 	at org.jenkinsci.test.acceptance.po.CapybaraPortingLayerImpl.waitFor(CapybaraPortingLayerImpl.java:144)
[2023-12-12T18:31:32.537Z] 	at plugins.FavoriteTest.shouldBeAbleToFavoriteAJob(FavoriteTest.java:31)
[2023-12-12T18:31:32.537Z] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[2023-12-12T18:31:32.537Z] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[2023-12-12T18:31:32.537Z] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[2023-12-12T18:31:32.537Z] 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[2023-12-12T18:31:32.537Z] 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
[2023-12-12T18:31:32.538Z] 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
[2023-12-12T18:31:32.538Z] 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
[2023-12-12T18:31:32.538Z] 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
[2023-12-12T18:31:32.538Z] 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
[2023-12-12T18:31:32.538Z] 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
[2023-12-12T18:31:32.538Z] 	at org.jenkinsci.test.acceptance.junit.WithPlugins$RuleImpl$1.evaluate(WithPlugins.java:160)
[2023-12-12T18:31:32.538Z] 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
[2023-12-12T18:31:32.538Z] 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
[2023-12-12T18:31:32.538Z] 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
[2023-12-12T18:31:32.538Z] 	at org.jenkinsci.test.acceptance.junit.JenkinsAcceptanceTestRule$1$1.evaluate(JenkinsAcceptanceTestRule.java:153)
[2023-12-12T18:31:32.538Z] 	at org.jenkinsci.test.acceptance.junit.FilterRule$1.evaluate(FilterRule.java:61)
[2023-12-12T18:31:32.538Z] 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
[2023-12-12T18:31:32.538Z] 	at org.jenkinsci.test.acceptance.junit.JenkinsAcceptanceTestRule$1.evaluate(JenkinsAcceptanceTestRule.java:53)
[2023-12-12T18:31:32.538Z] 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
[2023-12-12T18:31:32.538Z] 	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
[2023-12-12T18:31:32.538Z] 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
[2023-12-12T18:31:32.538Z] 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
[2023-12-12T18:31:32.538Z] 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
[2023-12-12T18:31:32.539Z] 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
[2023-12-12T18:31:32.539Z] 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
[2023-12-12T18:31:32.539Z] 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
[2023-12-12T18:31:32.539Z] 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
[2023-12-12T18:31:32.539Z] 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
[2023-12-12T18:31:32.539Z] 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
[2023-12-12T18:31:32.539Z] 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
[2023-12-12T18:31:32.539Z] 	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:316)
[2023-12-12T18:31:32.539Z] 	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:240)
[2023-12-12T18:31:32.539Z] 	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:214)
[2023-12-12T18:31:32.539Z] 	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:155)
[2023-12-12T18:31:32.539Z] 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
[2023-12-12T18:31:32.539Z] 	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
[2023-12-12T18:31:32.539Z] 	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
[2023-12-12T18:31:32.539Z] 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
[2023-12-12T18:31:32.539Z] Caused by: org.openqa.selenium.NoSuchElementException: Unable to locate By.cssSelector: #fav_my-project in http://127.0.0.1:46779/
[2023-12-12T18:31:32.539Z] 	at org.jenkinsci.test.acceptance.po.CapybaraPortingLayerImpl.find(CapybaraPortingLayerImpl.java:197)
[2023-12-12T18:31:32.539Z] 	at org.jenkinsci.test.acceptance.po.CapybaraPortingLayerImpl.lambda$waitFor$1(CapybaraPortingLayerImpl.java:144)
[2023-12-12T18:31:32.539Z] 	at org.jenkinsci.test.acceptance.junit.Wait$1.apply(Wait.java:123)
[2023-12-12T18:31:32.540Z] 	at org.openqa.selenium.support.ui.FluentWait.until(FluentWait.java:205)
[2023-12-12T18:31:32.540Z] 	... 41 more
```
ref: https://ci.jenkins.io/job/Core/job/acceptance-test-harness/job/master/1014/consoleText
test: https://github.com/jenkinsci/acceptance-test-harness/blob/3e414f016042aeb2846e8bb14badfc6b641aad45/src/test/java/plugins/FavoriteTest.java#L19

Would you mind taking a look, please @mawinter69 @cronik ?

**Comment by mawinter69** [comment]
https://github.com/jenkinsci/acceptance-test-harness/pull/1442
